content/courses/advanced-ada/parts/data_types/types_representation.rst

```diff
@@ -2754,6 +2761,402 @@ these array declarations are equivalent:
+package, to handle addresses. In contrast to other programming languages (such
+as C or C++), an address in Ada isn't an integer value |mdash| its definition
+actually depends on the compiler implementation, but for now, let's consider
+it to be a private type.
+
+The :ada:`Address` type has support for
+:ref:`address comparison <Adv_Ada_Address_Comparison>` and
+:ref:`address arithmetic <Adv_Ada_Address_Arithmetic>` (also
+known as *pointer arithmetic* in C). We discuss these topics later in this
+section. First, let's talk about the :ada:`Address` attribute and the
+:ada:`Address` aspect.
+
+.. admonition:: In the Ada Reference Manual
+
+    - :arm22:`13.7 The Package System <13-7>`
+
+
+.. _Adv_Ada_Address_Attribute:
+
+Address attribute
+~~~~~~~~~~~~~~~~~
+
+The :ada:`Address` attribute allows us to get the address of an object.
+For example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Type_Representation.Addresses.Address_Attribute
+
+    with System; use System;
+
+    procedure Use_Address is
+       I : aliased Integer := 5;
+       A : Address;
+    begin
+       A := I'Address;
+    end Use_Address;
+
+Here, we're assigning the address of the :ada:`I` object to the :ada:`A` address.
+
+.. admonition:: In the GNAT toolchain
+
+    GNAT offers a very useful extension to the :ada:`System` package to
+    retrieve a string for an address: :ada:`System.Address_Image`. This is the
+    function profile:
+
+    .. code-block:: ada
+
+        function System.Address_Image
+          (A : System.Address) return String;
+
+    We can use this function to display the address in an user message, for
+    example:
+
+    .. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Type_Representation.Addresses.Show_Address_Attribute
+
+        with Ada.Text_IO; use Ada.Text_IO;
+        with System.Address_Image;
+
+        procedure Show_Address_Attribute is
+           I  : aliased Integer := 5;
+        begin
+           Put_Line ("Address : "
+                     & System.Address_Image (I'Address));
+        end Show_Address_Attribute;
+
+.. admonition:: In the Ada Reference Manual
+
+    - :arm22:`13.3 Operational and Representation Attributes <13-3>`
+    - :arm22:`13.7 The Package System <13-7>`
+
+
+.. _Adv_Ada_Address_Aspect:
+
+Address aspect
+~~~~~~~~~~~~~~
+
+Usually, we let the compiler select the address of an object in memory, or let
+it use a register to store that object. However, we can specify the address of
+an object with the :ada:`Address` aspect. In this case, the compiler won't
+select an address automatically, but use the address that we're specifying. For
+example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Type_Representation.Addresses.Address_Aspect
+
+    with System; use System;
+    with System.Address_Image;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Address is
+
+       I_Main   : Integer;
```

Comment:
Should mark I_Main as aliased... but then the remaining examples below might look like they are incorrect.  Not sure what to really suggest. Maybe a note in prose?

Reply:
I've just marked `I_Main` as aliased. (I hope this doesn't become confusing for the readers.)